src/main.zig

```diff
@@ -53,12 +54,19 @@ fn init(io: *IO, cluster: u32, replica: u8, dir_fd: os.fd_t) !void {
     assert(filename.len == filename_len);
 
     // TODO Expose data file size on the CLI.
-    _ = try io.open_file(
+    const file = try io.open_file(
         dir_fd,
         filename,
-        config.journal_size_max, // TODO Double-check that we have space for redundant headers.
+        config.journal_size_max,
         true,
     );
+    defer std.os.close(file);
+
+    var sector: usize = 0;
+    while (sector * config.sector_size < config.journal_size_max) : (sector += 1) {
+        const write_size = try std.os.write(file, &Journal.format(cluster, sector));
```

Comment:
> I think there might also be a bug in how we call `Journal.format()` passing sector as a multiple of 4 KiB bytes, whereas `Journal.format()` internally seems to use sector not as a bytes value but as a sector index?

Yes, `Journal.format()` uses sector as an index, not a byte offset. But I think that is what is being passed in as well.

The condition does use bytes (`sector * config.sector_size < config.journal_size_max`) but it looks correct to me.

Reply:
> Perhaps we should rework the formatting slightly, to also take a larger buffer provided as a `target` slice, so that it can do the format in terms of larger writes (e.g. 4 MiB), to make the format significantly faster at the CLI for the user.

Instead of changing the interface, what if the caller just used an `std.io.BufferedWriter`?

Reply:
> The loop condition does use bytes (sector * config.sector_size < config.journal_size_max) but it looks correct to me.

Ah, yes. My eyes only saw the loop condition using bytes and so I thought that corresponded to the unit passed to `Journal.format()`.

Reply:
> Instead of changing the interface, what if the caller just used an std.io.BufferedWriter?

`std.io.BufferedWriter` would bring in too much machinery and abstraction. We don't use the std lib for I/O typically (with the exception of io_uring that we contributed) so that we have more visibility into exactly what's happening.

If possible, `Journal.format()` can assert that the `target` buffer is always fully used, so that there's no issue with padding.

Another idea is to have `Journal` take care of the writing when formatting, by using `Storage`. This would then wrap up all the logic related to formatting the `Journal` zone within `Journal`.

Reply:
For the superblock zone we have `SuperBlock.format()` which uses Storage and takes a callback: https://github.com/coilhq/tigerbeetle/blob/lsm-trees-and-more/src/lsm/superblock.zig#L489-L494

Agreed that having `Journal.format()` take `*Storage` and a callback, and call `write_sectors()` internally is not as simple.

We can always keep it as you have it, but with a `target` buffer instead of returning a sector so that we can use a 4 MiB buffer to avoid the performance showstopper. Then we can let it simmer some more, and see if it should take a storage pointer and callback just like the superblock zone.